examples/ssao/shaders/ssaoFragment.glsl

```diff
@@ -0,0 +1,282 @@
+    return vec2(cos(angle), sin(angle));
+}
+
+int getMipLevel(float ssR) {
+    return int(clamp(floor(log2(ssR)) - float(LOG_MAX_OFFSET), 0.0, float(MAX_MIP_LEVEL)));
+}
+
+/**
+ * Reconstructs tap location to a view-space point.
+ * This version uses several mipmaps bound as texture units
+ */
+vec3 getOffsetedPixelPos(ivec2 ssC, vec2 unitOffset, float screenSpaceRadius) {
+
+    int mipLevel = getMipLevel(screenSpaceRadius);
+
+    ivec2 ssP = ivec2(screenSpaceRadius * unitOffset) + ssC;
+    vec2 ssP_float = vec2(ssP);
+
+    float d = 0.0;
+    if (mipLevel == 1)
+        d = decodeFloatRGBA(texture2D(uMipmap0, (ssP_float + vec2(0.5)) / RenderSize));
+    else if (mipLevel == 2)
+        d = decodeFloatRGBA(texture2D(uMipmap1, (ssP_float + vec2(0.5)) / RenderSize));
+    else if (mipLevel == 3)
+        d = decodeFloatRGBA(texture2D(uMipmap2, (ssP_float + vec2(0.5)) / RenderSize));
+    else if (mipLevel == 4)
+        d = decodeFloatRGBA(texture2D(uMipmap3, (ssP_float + vec2(0.5)) / RenderSize));
+    else if (mipLevel == 5)
+        d = decodeFloatRGBA(texture2D(uMipmap4, (ssP_float + vec2(0.5)) / RenderSize));
+    else
+        d = decodeFloatRGBA(texture2D(uDepthTexture, (ssP_float + vec2(0.5)) / RenderSize));
+
+    return reconstructCSPosition(ssP_float + vec2(0.5), uNear + (uFar - uNear) * d);
+}
+
+/**
+ * Reconstructs tap location to a view-space point.
+ * This version uses several mipmaps generated automatically by the call glGenerateMipmaps
+ */
+// vec3 getOffsetedPixelPos(ivec2 ssC, vec2 unitOffset, float screenSpaceRadius) {
+
+//     int mipLevel = getMipLevel(screenSpaceRadius);
+
+//     ivec2 ssP = ivec2(screenSpaceRadius * unitOffset) + ssC;
+//     vec2 ssP_float = vec2(ssP);
+
+//     float d = decodeFloatRGBA(texture2DLodEXT(uDepthTexture, (ssP_float + vec2(0.25)) / RenderSize, float(mipLevel)));
+
+//     vec3 P;
+//     P.z = uNear + (uFar - uNear) * d;
+//     // Offset to pixel center
+//     P = reconstructCSPosition((ssP_float + vec2(0.5)), P.z);
+
+//     // DEBUG
+//     // if (mipLevel == 3)
+//     //     return reconstructCSPosition((vec2(ssP) + vec2(0.5)), 0.0);
+//     // END DEBUG
+//     return P;
+// }
+
+/**
+ * Reconstructs tap location to a view-space point.
+ * This version does not use mipmap levels.
+ */
+// vec3 getOffsetedPixelPos(ivec2 ssC, vec2 unitOffset, float screenSpaceRadius) {
+
+//     ivec2 ssP = ivec2(screenSpaceRadius * unitOffset) + ssC;
+//     vec2 ssP_float = vec2(ssP);
+
+//     return reconstructCSPosition(vec2(ssP) + vec2(0.5), zValueFromScreenSpacePosition(ssP_float));
+// }
+
+/**
+ * Computes the AO contribution for a single tap.
+ * The HIGH_QUALITY macro allows to smooth the AO
+ * and spread darken corners.
+ */
+float fallOffMethod0(float vv, float vn, vec3 normal) {
+
+    #   if HIGH_QUALITY
+
+        float invRadius2 = 1.0 / uRadius2;
+        float f = max(1.0 - vv * invRadius2, 0.0);
+        return f * max((vn - uBias) * inversesqrt(EPSILON + vv), 0.0);
```

Comment:
yes because :

```
vec3 v = occludingPoint - cameraSpacePosition;
float vv = dot(v, v);
float vn = dot(v, normal);
```

v is a view space vector so both vv and vn are dependent of the scene size (near/far).
An epsilon of 0.001 for a small model breaks the ssao easily for example.